### DISCUSSION TITLE
Ability to define optional material in a Course

### DISCUSSION THREAD

**Discussion by mmanimegalai** [Sunbird-Ed/Community#34] [Feature request]
This is a DIKSHA request

Currently all the course materials added to a course are mandatory. Completion of each course material is calculated as part of the Course Progress. However, the courses very often contain certain reference materials. Users taking the course can optionally use them and is not mandatory to consume it. Since all the material is mandatory now, users are forced to consume the reference materials too, in order to complete the course.

To avoid the above problem, following is the new functionality required for the Courses:
1. Courses support adding materials that are optional.
2. Consumption of the optional materials is not calculated as part of the course progress.
3. Users have a clear indication of the materials that are optional.

**Comment by pramodkvarma** [comment]
This is a popular request and I think internal APIs may already support it. @vrayulu ?

**Reply by vrayulu** [reply]
We have some APIs available for part of this capability. A few more APIs have to be updated/implemented to fully enable this.

1. We have APIs to create courses (any collection) with some content marked as optional within the course (Sunbird Knowlg). 
2. Progress tracking APIs assume all content in a collection are mandatory. This needs to be updated to handle optional content (Sunbird Lern).

In addition to the progress tracking APIs, creation and consumption UIs need to be enhanced.

**Comment by luckynvdu** [comment]
@mmanimegalai @pramodkvarma @vrayulu We have the following Jira tickets for the same -

ED - https://project-sunbird.atlassian.net/browse/SB-23313
Lern - https://project-sunbird.atlassian.net/browse/SB-26582,  https://project-sunbird.atlassian.net/browse/SB-28326

@rhwarrier request you to please prioritize these tickets for the request above

@sharathkashyap

**Reply by rhwarrier** [reply]
@reshmi-nair let's move these tickets to the new SB Lern project and line them up for 5.1

**Reply by sharathkashyap** [reply]
@mmanimegalai As the tickets has been identified and planned to be part of further release. I hereby request you to followup with the specific ticket further. Locking this thread from further conversation

**Comment by mmanimegalai** [comment]
Looping in @parthlawate

**Comment by parthlawate** [comment]
One more scenario of this that we have seen in the corporate learning side is scenarios like the one below 

1. Simple what you mentioned - Ability to Mark content as 'required' 
2. Ability to within a Sub Section in a course - Do any 2 any 3 (especially relevant when content are Quizzes)

**Reply by Krishnaj20** [reply]
@reshmi-nair @vinukumar-vs  - tagging you to share the tech design for optional material in courses in this thread for community feedback.

@mathewjpallan @rhwarrier

**Reply by reshmi-nair** [reply]
@parthlawate @Krishnaj20 Design on optional materials from Lern components view:
[Optional+Material+in+a+course-Design](https://project-sunbird.atlassian.net/wiki/spaces/SBDES/pages/3053322252/SB-26582+Optional+Material+in+a+course)
Please share your comments.
@rhwarrier @mathewpallan @vinukumar-vs

**Reply by Krishnaj20** [reply]
thanks @reshmi-nair .

@vinukumar-vs @swayangjit @rajeevsathish - Pls go thru the design and share your feedback.
@vrayulu @maheshkumargangula

**Reply by swayangjit** [reply]
@reshmi-nair In the backward-compatibility usecase this scenario might fail as old users can discover this course and consume it. The course completion will be hampered as no optional material-related code will be there for the old user base. 
cc: @vinukumar-vs @Krishnaj20

**Reply by kameshbhr** [reply]
I have added my comments to the confluence doc. @Krishnaj20 There are few things to be finalized/clarified in terms of the requirements too.

**Reply by parthlawate** [reply]
Not able to access the document .. please check access.. 
![Screenshot_20221223-071621](https://user-images.githubusercontent.com/673176/209262567-3aad70ff-70ef-46bf-a4dd-172a1b6b9b72.png)

**Comment by rhwarrier** [comment]
@swayangjit  can you please block some time tomorrow for the team to go over the design related queries with DC? The doc has already been reviewed by DC once, and we should meet and close open questions asap. We won't be able to deliver it as per plan if not.
@reshmi-nair  @vrayulu  @maheshkumargangula @rajeevsathish @Krishnaj20 @vinukumar-vs @kameshbhr

**Reply by Krishnaj20** [reply]
@pallakartheekreddy
